Improve CHECK TABLE system query

vdimir opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):

- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Query`CHECK TABLE` has better performance and usability (sends progress updates, cancellable)

### TODO

- [x] Document this query
- [x] Add test for progress and cancellation

vdimir committed:
Query CHECK TABLE takes care about progress and cancellation

robot-clickhouse-ci-2 commented:
<!-- automatic status comment for PR #52745 from ClickHouse/ClickHouse:vdimir/check_table_improvements -->
This is an automated comment for commit d515b6fb0ac2e5ba772652bbdab9d10b65faea4a with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52745/d515b6fb0ac2e5ba772652bbdab9d10b65faea4a/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

nikitamikhaylov commented:
I'm not sure, but I thought this query is single-threaded.. Is it true? Can we make it parallelized?

alesapin commented on src/Interpreters/InterpreterCheckQuery.cpp:
```diff
@@ -17,63 +18,158 @@ namespace DB
+        if (is_check_completed)
+            return {};
+
+        auto status = worker_result.wait_for(std::chrono::milliseconds(100));
```
Is it something common for progress (looks ad-hoc).

alesapin reviewed (commented):
I think we have tests for progress, maybe it make sense to add it here.

vdimir committed:
Build proper pipeline for CHECK TABLE

vdimir committed:
Upd tests for CHECK TABLE

vdimir committed:
Upd doc for CHECK TABLE

vdimir commented on src/Interpreters/InterpreterCheckQuery.cpp:
```diff
@@ -17,63 +18,158 @@ namespace DB
+
+                check_results.push_back(check_result);
+
+                bool should_continue = check_result.success || !check_query_single_value_result;
```
I assumed it was good to have a mode with early exit and thought we could do it since the user couldn't know how many errors there were with single-value result. However, its doubtful, so I returned it to the original behavior.

vdimir commented:
@alesapin 

I updated the implementation to build a more sophisticated pipeline and use its features properly.

It allowed to eliminate `std::async` and `sleep` entirely and execute everything in the pipeline executor pool. And also query is naturally parallelized like any other query.

I split `checkData` into two methods:
The first method obtains some container with a list of work to check table (parts or files depending on an engine). And the second executes one entry from this list.
We retrieve all sets of parts to check and then simultaneously execute entries from it by several processors.

I also added a doc (we have this query documented, so I added more information and examples) and added a test for progress.

alesapin commented on src/Interpreters/InterpreterCheckQuery.cpp:
```diff
@@ -8,72 +8,201 @@
+
+    Chunk generate() override
+    {
+        if (is_valuer_emitted.exchange(true))
```
we return chunk only once, right?

vdimir replied:
Yes, only one block with single row is expected in the result

alesapin commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -8487,24 +8486,30 @@ CheckResults StorageReplicatedMergeTree::checkData(const ASTPtr & query, Context
+{
 
-        for (auto & part : data_parts)
+    if (auto part = static_cast<DataValidationTasks *>(check_task_list.get())->next())
```
Looks a little bit scary to me, when we downcast using `static_cast`. Actually here it looks safe, but still  `dynamic_cast/assert_cast/typeid_cast` would be more natural here?

vdimir committed:
small fixes for CHECK TABLE

vdimir commented:
> Docs Check — Found errors in docs [Details](https://s3.amazonaws.com/clickhouse-test-reports/52745/ceaee4375df9c2189777a55b18a693fc847e54d3/docs_check.html)

```

[ERROR] Error: Docusaurus found broken links!

Please check the pages of your site in the list below, and make sure you don't reference any path that does not exist.
Note: it's possible to ignore broken links with the 'onBrokenLinks' Docusaurus configuration, and let the build pass.

Exhaustive list of all broken links found:

- On source page path = /docs/en/sql-reference/statements/check-table:
   -> linking to ../../operations/settings/settings.md#settings-max_threads (resolved as: /docs/en/operations/settings/settings.md)

- On source page path = /docs/zh/sql-reference/statements/check-table:
   -> linking to ../../operations/settings/settings.md#settings-max_threads (resolved as: /docs/zh/operations/settings/settings.md)

```

The path `../../operations/settings/settings.md#settings-max_threads` should be ok

```bash
$ cd docs/en/sql-reference/statements
$ cat ../../operations/settings/settings.md | grep -F settings-max_threads
## max_threads {#settings-max_threads}
Parallel `INSERT SELECT` has effect only if the `SELECT` part is executed in parallel, see [max_threads](#settings-max_threads) setting.
$ 
```

vdimir committed:
fix doc links

vdimir committed:
fix doc links

tavplubix commented:
Looks like it makes `01710_projection_part_check` flaky:
https://s3.amazonaws.com/clickhouse-test-reports/0/57025eed65ab0962464c23254c3394ecbfd98aef/stateless_tests__asan__[4_4].html
https://s3.amazonaws.com/clickhouse-test-reports/0/57025eed65ab0962464c23254c3394ecbfd98aef/stateless_tests__msan__[2_6].html

tavplubix commented:
And the CI caught it, it failed on commit dc98d3393804590b3c6aab6a1c049e14d345b41b:  
https://s3.amazonaws.com/clickhouse-test-reports/52745/dc98d3393804590b3c6aab6a1c049e14d345b41b/fast_test.html

![Screenshot from 2023-08-10 13-43-31](https://github.com/ClickHouse/ClickHouse/assets/14847450/ab63ff2c-e51d-4b60-a39c-1769dd74fea6)
